### ISSUE TITLE
Blaze Client keeps half-closed connection after server sends FIN-ACK through an idle connection

### ISSUE THREAD

**Issue by RafalSumislawski** [http4s/http4s#4798]
This is a problem because both sides can't release resources allocated to the connection, but also, because blaze will later try and fail to use that connection.

I prepared code that reproduces this behaviour using `BlazeServer`'s `idleTimeout` as the mechanism closing the connection on server's side. 

The scenario:
- make a request 
- wait until server closes the connection (due to 5s idle timeout)
- make another request

For comparison there's the same scenario but using `AsyncHttpClient`. The behaviour of `AsyncHttpClient` is in my opinion the expected behaviour (It's also consistent with for example Firefox's behaviour).

```
  val server = BlazeServerBuilder[IO](munitExecutionContext)
    .withIdleTimeout(5.second)
    .bindLocal(12345)
    .withHttpApp(Router("/" -> HttpRoutes.of[IO] { case GET -> Root => IO.sleep(1.seconds) >> Ok("") }).orNotFound)
    .resource

  test("Blaze Client keeps half-closed connection after server sends FIN-ACK through an idle connection") {
    (
      server,
      BlazeClientBuilder[IO](munitExecutionContext).resource
    ).tupled.use { case (_, client) =>for {
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
      _ <- IO.sleep(6.seconds)
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
    } yield ()
    }.unsafeRunSync()
  }

  test("AsyncHttpClient closes connection after server sends FIN-ACK through an idle connection") {
    (
      server,
      AsyncHttpClient.resource[IO]()
    ).tupled.use { case (_, client) => for {
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
      _ <- IO.sleep(6.seconds)
      _ <- client.expect[String](Uri.fromString(s"http://127.0.0.1:12345").yolo).attemptTap(result => IO(println(result)))
    } yield ()
    }.unsafeRunSync()
  }
```
(Full runnable code: https://github.com/RafalSumislawski/http4s/tree/blaze-half-closed-connection-issue)

The code itself just reproduces the behaviour, without checking if it's correct. I use wireshark to analyse to communication between the client and server.

### Blaze client:
<img width="839" alt="image" src="https://user-images.githubusercontent.com/9465998/116707900-77502e80-a9cf-11eb-83bf-8e77f96237d2.png">

In packet 1519 server closes the connection. The client ACKs it, but it doesn't send a FIN, so the connection remains half-closed (CLOSE_WAIT from server's perspective). Then packet 1521 is clients attempt to use that connection to send another request, to which the server responds with RST finally killing the connection. The client then establishes a new connection and successfully resends the request. Nonetheless precious resources have been wasted on the attempt to send request though closed connection. The retry mechanism was useful in this scenario, but it just covers the problem, not fixes it. And I see some other scenarios where it causes problems.

### AsyncHttpClient:
<img width="823" alt="image" src="https://user-images.githubusercontent.com/9465998/116708059-a070bf00-a9cf-11eb-8ede-7c834141f876.png">

After receiving FIN-ACK, the AsyncHttpClient immediately responds with FIN-ACK releasing resource on both ends. When a new request needs to be sent, I understands that it needs to open a new connection.

I realise this may be an issue inside blaze, not http4s, nonetheless I think the perspective of http4s-client user, and the comparison with other implementation is useful for clear definition of the problem and the expected behaviour.

**Comment by rossabaker** [comment]
My guess is that the problem is in blaze, but the AHC behavior looks better here.

**Comment by RafalSumislawski** [comment]
I'll try to find the source of the problem.

**Comment by RaasAhsan** [comment]
If you have the time and willingness to check, what is Ember Client's behavior here?

**Comment by RaasAhsan** [comment]
two questions about the network interactions here:
1. In packet 1519, the server is initiating a connection termination, so it sends a FIN, but it's also sending an ACK. Is that part of the termination handshake or is it ACKing something else?
2. If a JVM process observes a FIN from the peer, under what conditions does it reciprocate with a FIN? These connections are shuffled back into a pool, and AFAIK there's no way for us to be "notified" of the termination until we try to read or write on the socket again, which is what's happening here. This feels like something the OS should handle for us, but the fact that it works with AHC but not Blaze seems to suggest that Blaze is doing something wrong

**Comment by RafalSumislawski** [comment]
1. The ACK i packet 1519 is not part of the connection termination. In my opinion it is redundant, in the sense that it acknowledges reception of 0 bytes.

2. I have yet to understand how this works. My current idea is that we could keep reading from the connection while it's idle. We don't expect to read anything, but the read should fail with EOF when the connection terminates.

### Ember

It does this:
<img width="2" alt="image" src="https://user-images.githubusercontent.com/9465998/117185928-a6550e80-adda-11eb-9a25-ad345c16db38.png">

and throws this:
```
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:197)
	at sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishRead(UnixAsynchronousSocketChannelImpl.java:388)
	at sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(UnixAsynchronousSocketChannelImpl.java:191)
	at sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(UnixAsynchronousSocketChannelImpl.java:213)
	at sun.nio.ch.KQueuePort$EventHandlerTask.run(KQueuePort.java:301)
	at sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at async @ fs2.io.package$.asyncYield(io.scala:245)
	at flatMap @ fs2.io.tcp.SocketGroup.$anonfun$apply$8(SocketGroup.scala:291)
	at flatMap @ fs2.io.tcp.SocketGroup.getBufferOf$1(SocketGroup.scala:260)
	at flatMap @ fs2.io.tcp.SocketGroup.fs2$io$tcp$SocketGroup$$read0$1(SocketGroup.scala:290)
	at apply @ fs2.io.tcp.SocketGroup.apply(SocketGroup.scala:232)
	at unsafeRunSync @ org.http4s.client.blaze.BlazeIssues.$anonfun$new$1(BlazeIssues.scala:201)
	at withPermit @ fs2.io.tcp.SocketGroup.fs2$io$tcp$SocketGroup$$write0$1(SocketGroup.scala:339)
	at apply @ fs2.io.tcp.SocketGroup.apply(SocketGroup.scala:232)
	at withPermit @ fs2.io.tcp.SocketGroup.fs2$io$tcp$SocketGroup$$write0$1(SocketGroup.scala:339)
	at flatMap @ org.http4s.ember.core.Parser$Response$RespPrelude$.parsePrelude(Parser.scala:439)
	at flatMap @ org.http4s.ember.core.Parser$Response$.parser(Parser.scala:401)
```	

### JettyClient

Works nicely, just like AHC

<img width="831" alt="image" src="https://user-images.githubusercontent.com/9465998/117187041-c76a2f00-addb-11eb-9c17-47011505a4b8.png">

**Comment by RaasAhsan** [comment]
Interesting, so Ember doesn't even attempt to open a new connection? That's probably a blind spot in the pooling we should fix.

> I have yet to understand how this works. My current idea is that we could keep reading from the connection while it's idle. We don't expect to read anything, but the read should fail with EOF when the connection terminates.

I came across an article which seems to suggest reading a -1 will send back a FIN: https://mail.openjdk.java.net/pipermail/net-dev/2016-July/010088.html .

> Typically once a -1 is read by B, B will shut down the read half and 
will finish up writing any last bits of data on the socket, and then 
initiate its own close, shutting down the rest of the socket, causing B 
to transition from CLOSE_WAIT to LAST_ACK.  This causes B to send a FIN 
back to A, who then receives it and transitions to TIME_WAIT.

I wonder if Jetty and AHC are preemptively buffering a read?

**Comment by RafalSumislawski** [comment]
I checked jetty (using debuger). It keeps reading while connection is idle. It reads the -1 (EOF). Reading it does not result in sending FIN. The FIN is sent when Jetty calls `java.nio.channels.SocketChannel#close` (https://github.com/eclipse/jetty.project/blob/jetty-9.4.40.v20210413/jetty-io/src/main/java/org/eclipse/jetty/io/ChannelEndPoint.java#L210).

**Comment by RafalSumislawski** [comment]
I've opened a PR fixing this for blaze. I will open a separate issue for ember, so that we don't keep issues half-open (like ember keeps connections half-open ;) )

> Interesting, so Ember doesn't even attempt to open a new connection? That's probably a blind spot in the pooling we should fix.

I actually think that this retry behavior of blaze is more harmful than helpful. Sure it helps in this specific case. But in general, sending the same HTTP request twice is not safe. http4s doesn't have enough knowledge to say if the request is idempotent or not, or to say if the server started processing the request blaze sent in the first attempt. (see the issue no 2 I described in #4796). Furthermore, blazes retry it unbounded in number of attempts and time. A big no-no for reliability.
